nesta/core/routines/datasets/nih/nih_impute_base_id.py

```diff
@@ -0,0 +1,58 @@
+'''
+Impute Base ID
+==============
+
+What NiH don't tell you is that the `core_project_num` field
+has a base project number, which is effectively the actual
+core project number. This is essential for aggregating projects,
+otherwise it will appear that many duplicates exist in the data.
+
+This task imputes these values using a simple regex of the form:
+
+    {BASE_ID}-{an integer}-{an integer}-{an integer}
+
+Any `core_project_num` failing this regex are ignored.
+'''
+
+import logging
+import luigi
+from datetime import datetime as dt
+from multiprocessing.dummy import Pool as ThreadPool
+
+from nesta.core.luigihacks.mysqldb import make_mysql_target
+from nesta.packages.nih.impute_base_id import retrieve_id_ranges
+from nesta.packages.nih.impute_base_id import impute_base_id_thread
+
+
+class ImputeBaseIDTask(luigi.Task):
+    '''Impute the base ID using a regex of the form
+    {BASE_ID}-{an integer}-{an integer}-{an integer}
+
+    Args:
+        date (datetime): Date stamp.
+        test (bool): Running in test mode?
+    '''
+    date = luigi.DateParameter()
+    test = luigi.BoolParameter()
+
+    def output(self):
+        return make_mysql_target(self)
+
+    def run(self):
+        database = 'dev' if self.test else 'production'
+        id_ranges = retrieve_id_ranges(database)
+        # Threading required because it takes 2-3 days to impute
+        # all values on a single thread, or a few hours on 15 threads
+        pool = ThreadPool(15)
+        results = pool.map(impute_base_id_thread, id_ranges)
```

Comment:
`pool.starmap(impute_base_id_thread, id_ranges, database)` stays true to naming

Reply:
I'm not sure if this is how `starmap` works - from what I can see, by swapping from `map` to `starmap` I would keep everything the same (including the ugly repeated `database` arg) except I would change:

Reply:
since the pattern for `starmap` is:

```python
pool.starmap(func, iterable[, chunksize])
```

which is the same as `map`. Or have I missed something?

Reply:
You haven't missed something, my brain short-circuited.
I've often used `starmap` in combination with `itertools.product`, and ended up merging the two into one step.

What that would have suggested was:
```
from itertools import product
pool.starmap(impute_based_id_thread, product(id_ranges, database))
```

But the following is simpler (given `database` only ever takes the same value)
```
from itertools import repeat
pool.starmap(impute_based_id_thread, zip(id_ranges, repeat(database)))
```

Then you'd have `def impute_base_id_thread(id_range, database)`.

Reply:
actually in the end this doesn't work either as you need to unpack and then chain the `id_ranges` which have already been generated. One solution is:

```python
from itertools import chain

_id_ranges = map(lambda x: chain(x, [database]), id_ranges)
pool.starmap(impute_base_id_thread, _id_ranges)
```

which works nicely and is also purely iterable. I looked into using `partial` rather than the lambda, but it won't work since with `partial` unnamed args are parsed first and the `database` arg in my `chain` has to go last. No biggy, the lambda doesn't look too bad!

Reply:
Sorry to continue discussing this but I like little exercises like this :sweat_smile: 

To clarify I was still suggesting unpacking the range within the function (as below).
The suggested change to starmap was mostly about a clearer mapping between function/variable naming and what the code did: `retrieve_id_ranges` returns a range of id's (without added information about `database`); and the variable `id_ranges` were actually id ranges rather than id ranges and a `database` label.

```
from itertools import repeat
import multiprocessing.dummy as mpd

def foo(id_range, database):
    start, end = id_range
    print(f"{start}-{end}:{database}")

# Dummy data
database = "laterbase"
id_ranges = [(0, 2), (2, 3), (3, 3_00), (4, 5)]

pool = mpd.Pool()
pool.starmap(foo, zip(id_ranges, repeat(database)))

# OUTPUT
0-2:laterbase
2-3:laterbase
3-300:laterbase
4-5:laterbase
```

PS. not saying it should be changed or anything, just wanted to make my intentions clearer

Reply:
Gotcha, thanks!